[PHPUnit bridge] Add documentation for the component

theofidry opened the pull request:
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | #5059 |

First draft of the doc. I'm not hyper familiar with the component so I may have overlooked a feature. Also I kind of find the name off: _PHPUnit Bridge_, it's not about a bridge, it's about deprecation. Bridge is the technical solution to bring it, it does not reflect what the component does.

theofidry committed:
Add doc for the PHPUnit bridge component

theofidry committed:
Fix linting

theofidry committed:
Various fix

theofidry commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+
+It comes with the following features:
+
+* Enforce a consistent `C` locale (It forces applications to use the default language for output)
```
change for double ticks

javiereguiluz replied:
A don't understand the meaning of this: `Enforce a consistent `C` locale`

theofidry commented:
Any feedback on this one? @weaverryan? @javiereguiluz?

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+The PHPUnit Bridge Component
+=============================
+
+    The PHPUnit Bridge component provides utilities to report trigerred errors,
```
I'm not sure about the `trigerred errors` part. What does that mean exactly?

theofidry replied:
not sure how to reword that, could get rid of `trigerred` actually?

javiereguiluz replied:
Yes, I'd remove the `to report trigerred errors` part.

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+
+* Enforce a consistent `C` locale (It forces applications to use the default language for output)
+* Auto-register `class_exists` to load Doctrine annotations (when used);
+* Print a user deprecation notices summary at the end of the test suite;
```
`It displays the whole list of deprecated features used in the application.` ?

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+Usage
+-----
+
+Once the component installed, it automatically registers a
```
Let's simplify this:

```
Once the component installed, it automatically registers a
`PHPUnit event listener`_. This listener simply registers a `PHP error handler`_
called `DeprecationErrorHandler`. 
```

by this:


```
Once the component installed, it automatically registers a
`PHPUnit event listener`_ which in turn registers a `PHP error handler`_
called `DeprecationErrorHandler`. 
```

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+Once the component installed, it automatically registers a
+`PHPUnit event listener`_. This listener simply registers a `PHP error handler`_
+called `DeprecationErrorHandler`. After running your PHPUnit tests again, you
+will have a similar report:
```
`you will have a similar report` -> `you will get a report similar to this one:`

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+The summary includes:
+
+* **Unsilenced** reports deprecation notices that were triggered without the recommended @-silencing operator;
+* **Legacy** deprecation notices denote tests that explicitly test some legacy interfaces.
```
`some legacy interfaces` -> `some legacy features` ?

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+
+The summary includes:
+
+* **Unsilenced** reports deprecation notices that were triggered without the recommended @-silencing operator;
```
The lines of the three elements of this list are a bit long. Please wrap them to be less than 80 character in length.

theofidry replied:
how do you do that? If I do:

```
* Beginning of my line
second part
```

It will fire a warning:

```
WARNING: Bullet list ends without a blank line; unexpected unindent.
```

HeahDude replied:
you must keep the indent on every line :
```php
*   Beginning of my line
    second part
*   Another beginning

see http://www.sphinx-doc.org/en/stable/rest.html#lists-and-quote-like-blocks

javiereguiluz replied:
In lists, align the beginning of each item contents:

```
* Beginning of my line
  second part
```

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+Trigger deprecation notices
+---------------------------
+
+Deprecation notices can be triggered by using:
```
The PHP "hack" for RST docs require to have two colons instead of one: `using:` should be `using::`

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+Without the @-silencing operator, users would need to opt-out from deprecation
+notices. Silencing by default swaps this behavior and allows users to opt-in
+when they are ready to cope with them (by adding a custom error handler like the
+one provided by this bridge.) When not silenced, deprecation notices will appear
```
`this bridge.)` -> `this bridge).`

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+-------------
+
+In case you need to inspect the stack trace of a particular deprecation
+triggered by your unit tests, you can set the `SYMFONY_DEPRECATIONS_HELPER`
```
double backticks for `SYMFONY_DEPRECATIONS_HELPER`

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+In case you need to inspect the stack trace of a particular deprecation
+triggered by your unit tests, you can set the `SYMFONY_DEPRECATIONS_HELPER`
+`environment variable`_ to a regular expression that matches this deprecation's
+message, encapsed between `/`. For example, with:
```
`encapsed` sounds strange. MAybe use `wrapped` instead?

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+By default, any non-legacy-tagged or any non-@-silenced deprecation notices will
+make tests fail. Alternatively, setting `SYMFONY_DEPRECATIONS_HELPER` to the
+value `"weak"` will make the bridge ignore any deprecation notices. This is
+useful to projects that must use deprecated interfaces for backward compatibility
```
`deprecated interfaces` -> `deprecated features`

javiereguiluz commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+message contains the `"foobar"` string.
+
+By default, any non-legacy-tagged or any non-@-silenced deprecation notices will
+make tests fail. Alternatively, setting `SYMFONY_DEPRECATIONS_HELPER` to the
```
Double backticks for `SYMFONY_DEPRECATIONS_HELPER`

javiereguiluz commented:
@theofidry the contributed doc is nice. Thanks!

The only thing I miss is a cool feature that @nicolas-grekas told me about: you can use a number as the value of `SYMFONY_DEPRECATIONS_HELPER`. Why? @WouterJ explains it well in this StackOverfow answer: http://stackoverflow.com/questions/35377529/problems-using-symfony2-phpunit/35378446#35378446

theofidry commented:
>you can use a number as the value of `SYMFONY_DEPRECATIONS_HELPER`

Indeed quite a nice feature. I will address your comments shortly, thanks for the review :)

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+=============================
+
+    The PHPUnit Bridge component provides utilities to report trigerred errors,
+    legacy tests and usage of deprecated code.
```
It provides a lot more than just the deprecation helper:

 * A [bootstrap file](https://github.com/symfony/phpunit-bridge/blob/master/bootstrap.php) to configure the env correctly for testing (file is automatically loaded by Composer during tests)
 * A [ClockMock](https://github.com/symfony/phpunit-bridge/blob/master/ClockMock.php) for mocking time in time sensitive tests
 * A [SymfonyTestListener](https://github.com/symfony/phpunit-bridge/blob/master/SymfonyTestsListener.php) that (a) automatically registers the ClockMock if the test has the group `time-sensitive` and (b) comes with a Skipped feature to only test the previously skipped tests. I don't  think (b) is very usefull for others though (it's used on appveyor, where tests are run twice, onces with min settings and one with max, during the second run only skipped tests of the first run are executed).

Please note that the SymfonyTestListener is also automatically loaded.

theofidry replied:
Do we want to talk about it still? The bootstrap files allow to set the locale and the annotations (which is mentioned just below). As we do mention that it automatically registers an event listener, do we need to add that it's called `SymfonyTestListener`?

For the `ClockMock`, I couldn't find any doc on it. What is it supposed to do in the bridge?

WouterJ replied:
I think we should rewrote the component summary (these lines) a bit. The summary only talks about the deprecation utilities, but it provides other testing utilities as well. I think we should change it to something like:

```rst
    The PHPUnit Bridge provides testing utilities for PHPunit, including a deprecation
    helper, a helper to mock time and a bootstrap file to setup a consistent testing
    environment.
```

WouterJ replied:
Btw, the clock mock mocks the time functions (`time()`, `microtime()`, `sleep()` and `usleep()`). This way, you can test methods relying on time very accurately.

For instance. Assume you have some test like this:
```php
$stopwatch = new Stopwatch();

$stopwatch->start();
sleep(1);
$duration = $stopwatch->stop();

$this->assertEquals(1, $duration);
```
Without the clock mock, this can return in failure as `$duration` might be `1.000023` (for example). If you use the clock mock, this will always return exactly `1` (as `sleep()` will just do `startTime + 1` instead of really going to sleep).

To enable the clockmock, annotate your test with `@group time-sensitive` or call `ClockMock::register(true)` (before test) and `ClockMock::register(false)` (after test) yourself.

I think the feature is pretty great and worth documenting (it doesn't have to be included in this PR though).

theofidry replied:
>The summary only talks about the deprecation utilities, but it provides other testing utilities as well

Well before you raised the point on the `ClockMock` I though that was the case to be fair...

Do we want to mention the bootstrap file? It's more an implementation detail than anything else right? So should we really talk about it in the introduction?

WouterJ replied:
Yeah, we can maybe skip it in the introduction. It's usefull to know what it does, but you already mentioned that just below this summary.

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,129 @@
+
+            <php>
+                <server name="KERNEL_DIR" value="app/" />
+                <env name="SYMFONY_DEPRECATIONS_HELPER" value="/foobar/" />
```
let's only show the PHP tag + env tag. There is no need to show the complete file

theofidry committed:
Update

theofidry committed:
minor fixes

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+   single: PHPUnitBridge
+   single: Components; PHPUnitBridge
+
+The PHPUnit Bridge Component
```
I would omit the component part here and below.

WouterJ replied:
You seem to have forgotten to change this

theofidry replied:
Ha right, although maybe it would be better to keep it to me more consistent:

* The DomCrawler Component
* The EventDispatcher Component
* The Filesystem Component
* ...

WouterJ replied:
Well, this is a bridge and not a component. But we don't have a bridges section on the docs though, so the components section is the best place to put it now.

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+  locale to ``C``)
+* Auto-register ``class_exists`` to load Doctrine annotations (when used)
+* It displays the whole list of deprecated features used in the application
+* Display the stack trace of a deprecation on-demand.
```
Displays

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+* **Remaining/Other** deprecation notices are all other (non-legacy) notices,
+  grouped by message, test class and method
+
+Trigger deprecation notices
```
Trigger Deprecation Notices

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+one provided by this bridge). When not silenced, deprecation notices will appear
+in the **Unsilenced** section of the deprecation report.
+
+Mark tests as legacy
```
Mark Tests as Legacy

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+In case you need to inspect the stack trace of a particular deprecation
+triggered by your unit tests, you can set the ``SYMFONY_DEPRECATIONS_HELPER``
+`environment variable`_ to a regular expression that matches this deprecation's
+message, encapsed between ``/``. For example, with:
```
[...] enclosed with `/` [...]

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+It comes with the following features:
+
+* Forces the application to use the default language locale for output (set the
+  locale to ``C``)
```
What about changing this to: "Forces the tests to use a consistent locale (it's set to ``C``)"

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+
+Once the component installed, it automatically registers a
+`PHPUnit event listener`_ which in turn registers a `PHP error handler`_
+called ``DeprecationErrorHandler``. After running your PHPUnit tests again, you
```
I think we can remove "again" here

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+* **Legacy** deprecation notices denote tests that explicitly test some legacy
+  features
+* **Remaining/Other** deprecation notices are all other (non-legacy) notices,
+  grouped by message, test class and method
```
let's make this a definition list:

```rst
**Unsilenced**
    Reports deprecation notices that were triggered without
    the recommended ``@``-silencing operator.
**Legacy**
    Shows the number of deprecation notices triggered in
    legacy tests.
**Remaining/Other**
    Lists all deprecation notices, grouped by message, test
    class and method.
```

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+* Make its class start with the ``Legacy`` prefix
+* Make its method start with ``testLegacy``
+* Make its data provider start with ``provideLegacy`` or ``getLegacy``
+* Add the ``@group legacy`` annotation to its class or method
```
this is the most recommended way, let's put it as the first item (and maybe add `(**recommended**)`)

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+
+        <?xml version="1.0" encoding="UTF-8"?>
+
+        <!-- http://phpunit.de/manual/4.1/en/appendixes.configuration.html -->
```
should be the very first line in the code block

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+`environment variable`_ to a regular expression that matches this deprecation's
+message, encapsed between ``/``. For example, with:
+
+.. configuration-block::
```
YOu can remove this directive (it's only used when showing multiple formats of the same config, but PHPunit only supports XML)

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+                 xsi:noNamespaceSchemaLocation="http://schema.phpunit.de/4.1/phpunit.xsd"
+                 backupGlobals="false"
+                 colors="true"
+                 bootstrap="app/autoload.php"
```
these attributes can also be removed

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,114 @@
+PHPUnit_ will stop your test suite once a deprecation notice is triggered whose
+message contains the ``"foobar"`` string.
+
+By default, any non-legacy-tagged or any non-@-silenced deprecation notices will
```
let's put this in a seperate (sub-)section. Making tests pass with deprecation notices is a very common question, so I want the answer to be easily findable on google.

theofidry committed:
Add SYMFONY_DEPRECATIONS_HELPER feature

theofidry committed:
Address some comments

theofidry commented:
I updated most of the comments, my apologies if I forgot one

theofidry committed:
minor fixes

WouterJ commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,115 @@
+    <phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+             xsi:noNamespaceSchemaLocation="http://schema.phpunit.de/4.1/phpunit.xsd"
+    >
+    <?xml version="1.0" encoding="UTF-8"?>
```
should be removed (or placed after the file comment)

WouterJ commented:
Cool, thanks for your work on this one! You seem to have misunderstood/forgotten 2 minor things, but we can fix those during the merge as well. :+1: from me!

theofidry commented:
You're welcome :smile:

theofidry committed:
Add link for @-silence operator

theofidry committed:
Drop 'Component' from the name

theofidry commented:
@WouterJ looks like the link for `@-silencing operator` did not work, any idea?

theofidry committed:
Add doc for the ClockMock

theofidry commented:
updated:

* Added a link for the `@ operator`
* Dropped the `Component` part in the name as it's a bridge and not a component
* Added doc for the `ClockMock`

WouterJ commented:
It looks great! Thanks for your quick response.

theofidry committed:
Fix typo in intro

theofidry commented:
Oh gosh I hate when GitHub collapses all comments even if they are know. Sorry I cannot find where you putted your comment:

>Yeah, we can maybe skip it in the introduction. It's usefull to know what it does, but you already mentioned that just below this summary.

Maybe I've just changed the intro to not mention `ClockMock`. I still like the idea of the intro rather than the bullet point list as the bullet point list already gives a fairly detailed list of the features, whereas the intro stays more vague and gives a bigger picture of what the bridge does.

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,146 @@
+make tests fail. Alternatively, setting ``SYMFONY_DEPRECATIONS_HELPER`` to an
+arbitrary value (ex: ``320``) will make the tests fails only if a higher number
+of deprecation notices is reached (``0`` is the default value). You can also set
+the value ``"weak"`` will make the bridge ignore any deprecation notices. This is
```
[...] which will [...]

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,146 @@
+Time-sensitive Tests
+--------------------
+
+The mock class ``ClockMock`` allows you to mock the time functions ``time()``,
```
The `ClassMock` class allows [...]

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,146 @@
+
+**Unsilenced**
+    Reports deprecation notices that were triggered without the recommended
+    `@-silencing operator`_.
```
I would add blank lines between the different items.

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,146 @@
+The PHPUnit Bridge
+==================
+
+    The PHPUnit Bridge component provides utilities to report legacy tests, and
```
I would remove "component" here.

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,146 @@
+
+It comes with the following features:
+
+* Forces the tests to use a consistent locale (``C``)
```
Can you add blank lines between the items and terminate all but the last item with a semicolon (use a full stop for the last item)? And can you do the same for the other lists?

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,146 @@
+calculate the duration time of your process, here 1 second. However, this test
+may fail because of the duration time might actually be `1.000023s`.
+
+To use the ``ClockMock`` in your test, you can:
```
[...] the `ClockMock` class in [...]

xabbuh commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,146 @@
+
+For example, assuming you have the following code:
+
+.. code-block:: php
```
If you terminate the previous paragraph with a double colon, you can omit the `code-block` directive here.

theofidry committed:
Fix @xabbuh comments

theofidry commented:
@xabbuh updated

xabbuh commented:
:+1:

javiereguiluz commented:
@theofidry thanks for writing this doc. I used it to write this blog post: http://symfony.com/blog/new-in-symfony-2-8-clock-mocking-and-time-sensitive-tests

The blog post was reviewed by Nicolas, so I'm afraid that the current pull request has some errors about how to enable/disable the clock mocking explicitly. In addition, we could mention the fact that clock mocking makes your tests execute much faster (because time passes instantly).

theofidry commented:
@javiereguiluz then I guess what is left is update the clock mocking part with your blog post ;) I will try to do that tonight but can't guarantee anything.

theofidry committed:
Enhance clock mocking doc

theofidry commented:
Updated :)

dupuchba commented on components/phpunit_bridge.rst:
```diff
@@ -0,0 +1,242 @@
+The PHPUnit Bridge
+==================
+
+    The PHPUnit Bridge provides utilities to report legacy tests, and usage of
```
I think it's better english to remove the "," after "legacy tests"

theofidry replied:
Fair, fixed

theofidry committed:
typo

theofidry commented:
Any update? Would love to see it merge this weekend :smile:

WouterJ commented:
Here you go, @theofidry. Thanks for the reminder, I completely forgot to merge this one.

I've done some final fixes in 0e56ff9. If you don't agree with them, feel free to comment :)

Thanks for this great PR, it's a very nice article!

theofidry commented:
Thanks @WouterJ, I'm ok with the final fixes ;)
